File: hrms/hr/page/attendance_records/attendance_records.js
frappe.pages['attendance-records'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __('Attendance Records'),
		single_column: true
	});

	let dashboard = new AttendanceDashboard(page);
	dashboard.show();
}

class AttendanceDashboard {
	constructor(page) {
		this.page = page;
		this.setup_filters();
		this.setup_layout();
	}

	show() {
		this.refresh();
	}

	setup_filters() {
		this.page.add_field({
			fieldname: 'date_range',
			label: __('Select date range'),
			fieldtype: 'DateRange',
			default: [frappe.datetime.month_start(), frappe.datetime.get_today()],
			change: () => this.refresh()
		});

		this.page.add_field({
			fieldname: 'employee',
			label: __('Employee'),
			fieldtype: 'Link',
			options: 'Employee',
			change: () => this.refresh()
		});

		this.page.add_field({
			fieldname: 'department',
			label: __('Department'),
			fieldtype: 'Link',
			options: 'Department',
			change: () => this.refresh()
		});
	}

	setup_layout() {
		this.container = $('<div class="attendance-dashboard">').appendTo(this.page.main);
		this.container.html(`
			<div class="metrics-row"></div>
			<div class="charts-row">
				<div class="chart-container" id="leave-bar-chart">
					<div class="chart-title">${__('Leave Days')}</div>
					<div class="chart-canvas"></div>
				</div>
				<div class="chart-container" id="wfh-pie-chart">
					<div class="chart-title">${__('WFH & Leave Breakdown')}</div>
					<div class="chart-canvas"></div>
				</div>
				<div class="chart-container" id="punctuality-pie-chart">
					<div class="chart-title">${__('Punctuality (On Time vs Late)')}</div>
					<div class="chart-canvas"></div>
				</div>
			</div>
			<div class="table-section">
				<table class="attendance-table">
					<thead>
						<tr>
							<th>${__('Date')}</th>
							<th>${__('Employee Name')}</th>
							<th>${__('In Time')}</th>
							<th>${__('Out Time')}</th>
							<th>${__('Duration')}</th>
							<th>${__('Status')}</th>
						</tr>
					</thead>
					<tbody></tbody>
				</table>
			</div>
		`);
	}

	refresh() {
		const dr = this.page.fields_dict.date_range.get_value();
		const filters = {
			from_date: dr ? dr[0] : null,
			to_date: dr ? dr[1] : null,
			employee: this.page.fields_dict.employee.get_value(),
			department: this.page.fields_dict.department.get_value()
		};

		frappe.call({
			method: "hrms.hr.page.attendance_records.attendance_records.get_dashboard_data",
			args: { filters: filters },
			callback: (r) => {
				if (r.message) {
					this.render_metrics(r.message.metrics);
					this.render_charts(r.message.charts);
					this.render_table(r.message.table_data);
				}
			}
		});
	}

	render_metrics(metrics) {
		const row = this.container.find('.metrics-row');
		row.empty();
		const items = [
			{ label: __('Total Time Worked'), value: metrics.total_time_worked },
			{ label: __('Avg. Check In'), value: metrics.avg_check_in },
			{ label: __('Avg. Check Out'), value: metrics.avg_check_out },
			{ label: __('Leave Taken'), value: metrics.leave_taken },
			{ label: __('WFH Days'), value: metrics.wfh_days }
		];

		items.forEach(item => {
			$(`
				<div class="metric-card">
					<div class="metric-value">${item.value}</div>
					<div class="metric-label">${item.label}</div>
				</div>
			`).appendTo(row);
		});
	}

	render_charts(charts) {
		this.render_bar("#leave-bar-chart .chart-canvas", charts.leave_summary);
		this.render_pie("#wfh-pie-chart .chart-canvas", charts.breakdown);
		this.render_pie("#punctuality-pie-chart .chart-canvas", charts.punctuality);
	}

	render_bar(selector, leave_summary) {
		if (!leave_summary || leave_summary.length === 0) {
			$(selector).html(`<div class="text-muted text-center">${__('No data')}</div>`);
			return;
		}

		const labels = leave_summary.map(d => d.leave_type);
		const taken = leave_summary.map(d => d.taken);
		const left = leave_summary.map(d => d.left);

		new frappe.Chart(selector, {
			data: {
				labels: labels,
				datasets: [
					{ name: __('Taken'), values: taken, chartType: 'bar' },
					{ name: __('Days Left'), values: left, chartType: 'bar' }
				]
			},
			type: 'bar',
			height: 250,
			colors: ['#0bc5ea', '#ed64a6'],
			barOptions: { stacked: 1 }
		});
	}

	render_pie(selector, breakdown) {
		const labels = Object.keys(breakdown);
		const values = Object.values(breakdown);

		if (values.length === 0 || values.every(v => v === 0)) {
			$(selector).html(`<div class="text-muted text-center">${__('No data')}</div>`);
			return;
		}

		new frappe.Chart(selector, {
			data: {
				labels: labels,
				datasets: [{ values: values }]
			},
			type: 'pie',
			height: 250,
			colors: ['#3182ce', '#38b2ac', '#9f7aea', '#ed64a6', '#f6ad55']
		});
	}

	render_table(data) {
		const tbody = this.container.find('.attendance-table tbody');
		tbody.empty();

		if (!data || data.length === 0) {
			tbody.append(`<tr><td colspan="6" class="text-center">${__('No records found')}</td></tr>`);
			return;
		}

		data.forEach(r => {
			const punctuality_class = r.late_entry ? 'late' : 'on-time';
			const punctuality_label = r.late_entry ? __('Late') : __('On Time');
			const in_time = r.in_time ? frappe.datetime.get_time(r.in_time) : '-';
			const out_time = r.out_time ? frappe.datetime.get_time(r.out_time) : '-';
			const duration = r.working_hours ? `${flt(r.working_hours, 2)} hrs` : '-';

			tbody.append(`
				<tr>
					<td>${frappe.datetime.str_to_user(r.attendance_date)}</td>
					<td>${r.employee_name}</td>
					<td>${in_time}</td>
					<td>${out_time}</td>
					<td>${duration}</td>
					<td><span class="punctuality-badge ${punctuality_class}">${punctuality_label}</span></td>
				</tr>
			`);
		});
	}
}

File: hrms/hr/page/attendance_records/attendance_records.py
import frappe
from frappe import _
from frappe.utils import flt, getdate, format_datetime, format_duration, get_time
from datetime import datetime, timedelta

@frappe.whitelist()
def get_dashboard_data(filters):
	if isinstance(filters, str):
		filters = frappe.parse_json(filters)
	
	from_date = filters.get("from_date")
	to_date = filters.get("to_date")
	employee = filters.get("employee")
	department = filters.get("department")
	
	conditions = {"docstatus": 1}
	if from_date:
		conditions["attendance_date"] = [">=", from_date]
	if to_date:
		if "attendance_date" in conditions:
			conditions["attendance_date"] = ["between", [from_date, to_date]]
		else:
			conditions["attendance_date"] = ["<=", to_date]
	if employee:
		conditions["employee"] = employee
	if department:
		conditions["department"] = department
	
	attendance_records = frappe.get_all("Attendance", 
		filters=conditions, 
		fields=["name", "employee", "employee_name", "attendance_date", "status", "in_time", "out_time", "working_hours", "late_entry", "early_exit", "department", "leave_type"],
		order_by="attendance_date desc"
	)
	
	metrics = calculate_metrics(attendance_records)
	charts = calculate_charts(attendance_records, employee, from_date, to_date)
	
	return {
		"metrics": metrics,
		"charts": charts,
		"table_data": attendance_records
	}

def calculate_metrics(records):
	total_time_worked_s = 0
	check_in_times = []
	check_out_times = []
	leave_taken = 0
	wfh_days = 0
	
	for r in records:
		if r.status in ["Present", "Half Day", "Work From Home"]:
			if r.working_hours:
				total_time_worked_s += flt(r.working_hours) * 3600
			
		if r.status == "On Leave":
			leave_taken += 1
		elif r.status == "Work From Home":
			wfh_days += 1
		elif r.status == "Half Day":
			leave_taken += 0.5
			
		if r.in_time:
			t = get_time(r.in_time)
			check_in_times.append(t.hour * 3600 + t.minute * 60 + t.second)
			
		if r.out_time:
			t = get_time(r.out_time)
			check_out_times.append(t.hour * 3600 + t.minute * 60 + t.second)
			
	avg_check_in = sum(check_in_times) / len(check_in_times) if check_in_times else 0
	avg_check_out = sum(check_out_times) / len(check_out_times) if check_out_times else 0
	
	return {
		"total_time_worked": format_duration(total_time_worked_s),
		"avg_check_in": format_seconds(avg_check_in),
		"avg_check_out": format_seconds(avg_check_out),
		"leave_taken": leave_taken,
		"wfh_days": wfh_days
	}

def format_seconds(seconds):
	h = int(seconds // 3600)
	m = int((seconds % 3600) // 60)
	s = int(seconds % 60)
	return f"{h:02}:{m:02}:{s:02}"

def calculate_charts(records, employee, from_date, to_date):
	# Breakdown
	breakdown = {}
	for r in records:
		if r.status == "Work From Home":
			breakdown["WFH"] = breakdown.get("WFH", 0) + 1
		elif r.status in ["On Leave", "Half Day"]:
			lt = r.leave_type or "Other"
			breakdown[lt] = breakdown.get(lt, 0) + (1 if r.status == "On Leave" else 0.5)
			
	# Punctuality
	punctuality = {"On Time": 0, "Late": 0}
	for r in records:
		if r.status in ["Present", "Half Day", "Work From Home"]:
			if r.late_entry:
				punctuality["Late"] += 1
			else:
				punctuality["On Time"] += 1
				
	# Leave Type Summary (for bar chart)
	leave_summary = []
	leave_types = frappe.get_all("Leave Type", fields=["name"])
	
	leave_allocation = {}
	if employee:
		from hrms.hr.doctype.leave_application.leave_application import get_leave_details
		from frappe.utils import nowdate
		res = get_leave_details(employee, nowdate())
		leave_allocation = res.get("leave_allocation", {})

	for lt in leave_types:
		details = leave_allocation.get(lt.name, {})
		taken = flt(details.get("leaves_taken", 0))
		balance = flt(details.get("remaining_leaves", 0))
		total = flt(details.get("total_leaves", 0))
		
		# If no allocation record, it won't be in leave_allocation
		# But we still want to show it if there are taken leaves in the selected period
		if lt.name in breakdown:
			period_taken = breakdown[lt.name]
			if not total: # ensure we show at least the taken amount if total is 0
				total = period_taken
		
		leave_summary.append({
			"leave_type": lt.name,
			"taken": taken,
			"left": balance,
			"total": total
		})
	
	# Handle leaves with no assigned type
	if breakdown.get("Other", 0) > 0:
		leave_summary.append({
			"leave_type": _("Other"),
			"taken": breakdown["Other"],
			"left": 0
		})
			
	return {
		"breakdown": breakdown,
		"punctuality": punctuality,
		"leave_summary": leave_summary
	}
